include/libassert/platform.hpp

```diff
@@ -0,0 +1,343 @@
+    #define LIBASSERT_ATTR_NOINLINE __declspec(noinline)
+    #define LIBASSERT_UNREACHABLE_CALL __assume(false)
+#endif
+
+#if LIBASSERT_IS_MSVC
+    #define LIBASSERT_STRONG_EXPECT(expr, value) (expr)
+#elif defined(__clang__) && __clang_major__ >= 11 || __GNUC__ >= 9
+    #define LIBASSERT_STRONG_EXPECT(expr, value) __builtin_expect_with_probability((expr), (value), 1)
+#else
+    #define LIBASSERT_STRONG_EXPECT(expr, value) __builtin_expect((expr), (value))
+#endif
+
+// deal with gcc shenanigans
+// at one point their std::string's move assignment was not noexcept even in c++17
+// https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html
+#if defined(_GLIBCXX_USE_CXX11_ABI)
+    // NOLINTNEXTLINE(misc-include-cleaner)
+    #define LIBASSERT_GCC_ISNT_STUPID _GLIBCXX_USE_CXX11_ABI
+#else
+    // assume others target new abi by default - homework
+    #define LIBASSERT_GCC_ISNT_STUPID 1
+#endif
+
+
+#if LIBASSERT_IS_GCC || LIBASSERT_STD_VER >= 20
+    // __VA_OPT__ needed for GCC, https://gcc.gnu.org/bugzilla/show_bug.cgi?id=44317
+    #define LIBASSERT_VA_ARGS(...) __VA_OPT__(,) __VA_ARGS__
+#else
+    // clang properly eats the comma with ##__VA_ARGS__
+    #define LIBASSERT_VA_ARGS(...) , ##__VA_ARGS__
+#endif
+
+
+///
+/// C++20 functionality wrappers.
+///
+
+// Check if we can use std::is_constant_evaluated.
+#ifdef __has_include
+    #if __has_include(<version>)
+        #include <version>
+        #ifdef __cpp_lib_is_constant_evaluated
+            #include <type_traits>
+            #define LIBASSERT_HAS_IS_CONSTANT_EVALUATED
+        #endif
+    #endif
+#endif
+
+
+// Check if we have the builtin __builtin_is_constant_evaluated.
+#ifdef __has_builtin
+    #if __has_builtin(__builtin_is_constant_evaluated)
+        #define LIBASSERT_HAS_BUILTIN_IS_CONSTANT_EVALUATED
+    #endif
+#endif
+
+
+// GCC 9.1+ and later has __builtin_is_constant_evaluated
+#if (__GNUC__ >= 9) && !defined(LIBASSERT_HAS_BUILTIN_IS_CONSTANT_EVALUATED)
+    #define LIBASSERT_HAS_BUILTIN_IS_CONSTANT_EVALUATED
+#endif
+
+
+// Visual Studio 2019 (19.25) and later supports __builtin_is_constant_evaluated
+#if defined(_MSC_FULL_VER) && (_MSC_FULL_VER >= 192528326)
+    #define LIBASSERT_HAS_BUILTIN_IS_CONSTANT_EVALUATED
+#endif
+
+
+namespace libassert::support
+{
+    /**
+     * @brief C++17 implementation of is_constant_evaluated that detects whether the function call occurs within a constant-evaluated context. Returns true if the evaluation of the call occurs within the evaluation of an expression or conversion that is manifestly constant-evaluated; otherwise returns false.
+     * @return true if the evaluation of the call occurs within the evaluation of an expression or conversion that is manifestly constant-evaluated; otherwise false.
+     *
+     * @note Works with C++17 under GCC 9.1+, Clang 9+, and MSVC 19.25.
+     */
+    constexpr bool is_constant_evaluated() noexcept
+    {
+#if defined(LIBASSERT_HAS_IS_CONSTANT_EVALUATED)
```

Comment:
Since preprocessor directives are being indented everywhere else they should be here too